nodeInterface/NodeInterface.go

```diff
@@ -413,6 +417,81 @@ func (n NodeInterface) ConstructOutboxProof(c ctx, evm mech, size, leaf uint64)
 	return send, root, hashes32, nil
 }
 
+func (n NodeInterface) GasEstimateComponents(
+	c ctx,
+	evm mech,
+	from, to addr,
+	ignoreTo bool,
+	gas uint64,
+	maxFeePerGas, maxPriorityFeePerGas, value huge,
+	data []byte,
+) (uint64, uint64, uint64, uint64, error) {
+
+	node, err := arbNodeFromNodeInterfaceBackend(n.backend)
+	if err != nil {
+		return 0, 0, 0, 0, err
+	}
+
+	context := n.context
+	chainid := evm.ChainConfig().ChainID
+	backend := node.Backend.APIBackend()
+	gasCap := backend.RPCGasCap()
+	block := rpc.BlockNumberOrHashWithNumber(rpc.BlockNumber(n.header.Number.Int64()))
+	nonce := uint64(0)
+
+	args := arbitrum.TransactionArgs{
+		ChainID:              (*hexutil.Big)(chainid),
+		From:                 &from,
+		Gas:                  (*hexutil.Uint64)(&gas),
+		MaxFeePerGas:         (*hexutil.Big)(maxFeePerGas),
+		MaxPriorityFeePerGas: (*hexutil.Big)(maxPriorityFeePerGas),
+		Value:                (*hexutil.Big)(value),
+		Nonce:                (*hexutil.Uint64)(&nonce),
+		Data:                 (*hexutil.Bytes)(&data),
+	}
+	if !ignoreTo {
+		args.To = &to
+	}
+
+	totalRaw, err := arbitrum.EstimateGas(context, backend, args, block, gasCap)
+	if err != nil {
+		return 0, 0, 0, 0, err
+	}
+	total := uint64(totalRaw)
+
+	pricing := c.State.L1PricingState()
+
+	msg, err := args.ToMessage(gasCap, n.header, evm.StateDB.(*state.StateDB))
+	if err != nil {
+		return 0, 0, 0, 0, err
+	}
+	poster, err := pricing.ReimbursableAggregatorForSender(from)
+	if err != nil {
+		return 0, 0, 0, 0, err
+	}
+	if poster == nil {
+		poster = &addr{}
+	}
+	feeForL1, _ := pricing.PosterDataCost(msg, from, *poster)
+
+	baseFee, err := c.State.L2PricingState().BaseFeeWei()
+	if err != nil {
+		return 0, 0, 0, 0, err
+	}
+	l1BaseFeeEstimate, err := pricing.L1BaseFeeEstimateWei()
+	if err != nil {
+		return 0, 0, 0, 0, err
+	}
+
+	// Compute the fee paid for L1 in L2 terms
+	//   See in GasChargingHook that this does not induce truncation error
+	//
+	feeForL1 = arbmath.BigMulByUfrac(feeForL1, 110, 100)
```

Comment:
We should also leave a comment in `GasChargingHook` that if the buffer ratio is changed there, it should be changed here too